File: app/src/main/java/com/thewizrd/simpleweather/main/WeatherRadarFragment.java
package com.thewizrd.simpleweather.main;

import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.ViewCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.transition.MaterialContainerTransform;
import com.google.android.material.transition.MaterialFadeThrough;
import com.thewizrd.common.controls.WeatherNowViewModel;
import com.thewizrd.shared_resources.utils.AnalyticsLogger;
import com.thewizrd.shared_resources.utils.ContextUtils;
import com.thewizrd.simpleweather.R;
import com.thewizrd.simpleweather.databinding.FragmentWeatherRadarBinding;
import com.thewizrd.simpleweather.fragments.ToolbarFragment;
import com.thewizrd.simpleweather.radar.RadarProvider;
import com.thewizrd.simpleweather.radar.RadarViewProvider;
import com.thewizrd.simpleweather.snackbar.SnackbarManager;

@RequiresApi(value = Build.VERSION_CODES.LOLLIPOP)
public class WeatherRadarFragment extends ToolbarFragment {
    private WeatherNowViewModel weatherView = null;
    private FragmentWeatherRadarBinding binding;

    private RadarViewProvider radarViewProvider;

    @NonNull
    @Override
    public SnackbarManager createSnackManager() {
        SnackbarManager mSnackMgr = new SnackbarManager(binding.getRoot());
        mSnackMgr.setSwipeDismissEnabled(true);
        mSnackMgr.setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_FADE);
        return mSnackMgr;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AnalyticsLogger.logEvent("WeatherRadarFragment: onCreate");
        setExitTransition(new MaterialFadeThrough());
        setEnterTransition(new MaterialFadeThrough());
        setSharedElementEnterTransition(new MaterialContainerTransform());
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ViewGroup root = (ViewGroup) super.onCreateView(inflater, container, savedInstanceState);
        // Use this to return your custom view for this Fragment
        binding = FragmentWeatherRadarBinding.inflate(inflater, root, true);

        ViewCompat.setTransitionName(binding.radarWebviewContainer, "radar");

        // Setup Actionbar
        getToolbar().setNavigationIcon(ContextUtils.getAttrResourceId(getToolbar().getContext(), R.attr.homeAsUpIndicator));
        getToolbar().setNavigationOnClickListener(v -> Navigation.findNavController(v).navigateUp());

        getToolbar().inflateMenu(R.menu.radar);
        getToolbar().setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.action_refresh:
                        if (radarViewProvider != null && weatherView.getLocationCoord() != null) {
                            radarViewProvider.updateCoordinates(weatherView.getLocationCoord(), true);
                        }
                        return true;
                }
                return false;
            }
        });

        radarViewProvider = RadarProvider.getRadarViewProvider(requireContext(), binding.radarWebviewContainer);
        radarViewProvider.enableInteractions(true);
        radarViewProvider.onCreateView(savedInstanceState);

        return root;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        weatherView = new ViewModelProvider(getAppCompatActivity()).get(WeatherNowViewModel.class);
        if (radarViewProvider != null && weatherView.getLocationCoord() != null) {
            radarViewProvider.onViewCreated(weatherView.getLocationCoord());
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        if (radarViewProvider != null) {
            radarViewProvider.onStart();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        AnalyticsLogger.logEvent("WeatherRadarFragment: onResume");

        if (radarViewProvider != null) {
            radarViewProvider.onResume();
        }
        initialize();
    }

    @Override
    public void onPause() {
        AnalyticsLogger.logEvent("WeatherRadarFragment: onPause");

        if (radarViewProvider != null) {
            radarViewProvider.onPause();
        }
        super.onPause();
    }

    @Override
    public void onStop() {
        if (radarViewProvider != null) {
            radarViewProvider.onStop();
        }
        super.onStop();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        if (radarViewProvider != null) {
            radarViewProvider.onSaveInstanceState(outState);
        }
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroyView() {
        if (radarViewProvider != null) {
            radarViewProvider.onDestroyView();
            radarViewProvider = null;
        }
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        if (radarViewProvider != null) {
            radarViewProvider.onLowMemory();
        }
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if (radarViewProvider != null) {
            radarViewProvider.onConfigurationChanged();
        }
    }

    @Override
    protected int getTitle() {
        return R.string.label_radar;
    }

    // Initialize views here
    @CallSuper
    protected void initialize() {
        if (radarViewProvider != null) {
            radarViewProvider.updateRadarView();
        }
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/main/WeatherRadarFragment.kt
package com.thewizrd.simpleweather.main

import android.app.Activity
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.RequiresApi
import androidx.appcompat.widget.Toolbar
import androidx.core.view.ViewCompat
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.transition.MaterialContainerTransform
import com.google.android.material.transition.MaterialFadeThrough
import com.thewizrd.shared_resources.utils.AnalyticsLogger
import com.thewizrd.shared_resources.utils.ContextUtils.getAttrResourceId
import com.thewizrd.simpleweather.R
import com.thewizrd.simpleweather.databinding.FragmentWeatherRadarBinding
import com.thewizrd.simpleweather.fragments.ToolbarFragment
import com.thewizrd.simpleweather.radar.RadarProvider.getRadarViewProvider
import com.thewizrd.simpleweather.radar.RadarViewProvider
import com.thewizrd.simpleweather.snackbar.SnackbarManager
import com.thewizrd.simpleweather.viewmodels.WeatherNowViewModel
import kotlinx.coroutines.launch

@RequiresApi(value = Build.VERSION_CODES.LOLLIPOP)
class WeatherRadarFragment : ToolbarFragment() {
    private val wNowViewModel: WeatherNowViewModel by activityViewModels()
    private lateinit var binding: FragmentWeatherRadarBinding

    private var radarViewProvider: RadarViewProvider? = null

    override fun createSnackManager(activity: Activity): SnackbarManager {
        val mSnackMgr = SnackbarManager(binding.root)
        mSnackMgr.setSwipeDismissEnabled(true)
        mSnackMgr.setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_FADE)
        return mSnackMgr
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AnalyticsLogger.logEvent("WeatherRadarFragment: onCreate")
        exitTransition = MaterialFadeThrough()
        enterTransition = MaterialFadeThrough()
        sharedElementEnterTransition = MaterialContainerTransform()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = super.onCreateView(inflater, container, savedInstanceState) as ViewGroup?
        // Use this to return your custom view for this Fragment
        binding = FragmentWeatherRadarBinding.inflate(inflater, root, true)

        ViewCompat.setTransitionName(binding.radarWebviewContainer, "radar")

        // Setup Actionbar
        toolbar.setNavigationIcon(toolbar.context.getAttrResourceId(R.attr.homeAsUpIndicator))
        toolbar.setNavigationOnClickListener { activity?.onBackPressedDispatcher?.onBackPressed() }

        toolbar.inflateMenu(R.menu.radar)
        toolbar.setOnMenuItemClickListener(Toolbar.OnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_refresh -> {
                    wNowViewModel.weather.value?.let {
                        radarViewProvider?.updateCoordinates(it.locationCoord, true)
                    }
                    return@OnMenuItemClickListener true
                }
            }
            false
        })

        radarViewProvider = getRadarViewProvider(requireContext(), binding.radarWebviewContainer)
        radarViewProvider!!.enableInteractions(true)
        radarViewProvider!!.onCreateView(savedInstanceState)

        return root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        wNowViewModel.weather.value?.let {
            radarViewProvider?.onViewCreated(it.locationCoord)
            radarViewProvider?.updateRadarView()
        }

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.RESUMED) {
                radarViewProvider?.updateRadarView()
            }
        }
    }

    override fun onStart() {
        super.onStart()
        radarViewProvider?.onStart()
    }

    override fun onResume() {
        super.onResume()
        AnalyticsLogger.logEvent("WeatherRadarFragment: onResume")

        radarViewProvider?.onResume()
    }

    override fun onPause() {
        AnalyticsLogger.logEvent("WeatherRadarFragment: onPause")

        radarViewProvider?.onPause()
        super.onPause()
    }

    override fun onStop() {
        radarViewProvider?.onStop()
        super.onStop()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        radarViewProvider?.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onDestroyView() {
        radarViewProvider?.onDestroyView()
        radarViewProvider = null

        super.onDestroyView()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        radarViewProvider?.onLowMemory()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        radarViewProvider?.onConfigurationChanged()
    }

    override val titleResId: Int
        get() = R.string.label_radar
}